From bbb310b90ecd34b3396c3ecda300264527bb211f Mon Sep 17 00:00:00 2001
Message-Id: <bbb310b90ecd34b3396c3ecda300264527bb211f.1440106296.git.chang-joon.lee@intel.com>
In-Reply-To: <1f426693c14c79e65ad8e62ef6bb4b07d6484040.1440106296.git.chang-joon.lee@intel.com>
References: <1f426693c14c79e65ad8e62ef6bb4b07d6484040.1440106296.git.chang-joon.lee@intel.com>
From: Tim Gore <tim.gore@intel.com>
Date: Tue, 14 Jul 2015 14:09:35 +0100
Subject: [PATCH 07/11] FOR_UPSTREAM [VPG]: drm/i915: clear STOP_RING_ALLOW_BAN

With a per engine reset (in i915_handle_hung_ring)
the STOP_RING_ALLOW_BAN and STOP_RING_ALLOW_ERROR bits
in gpu_error.stop_rings do not get cleared. This is
different behaviour from a full gpu reset and causes
some igt tests to fail because the igt_set_stop_rings
function requires stop_rings to be clear (ie 0) before
it will allow a new hang to be notified/injected.

So, update the i915_handle_hung_ring function to clear
these bits once all hung rings have been reset.

Tracked-On: https://jira01.devtools.intel.com/browse/VIZ-6030
Change-Id: Ib5d50c8456239b1b63e90838eab0509ea1379174
Signed-off-by: Tim Gore <tim.gore@intel.com>

## Don't forget to add a JIRA URL to the Tracked-On metadata!
## https://wiki.ith.intel.com/display/CACTUS/Patch+requirements+check
## url should be in the form:
## Tracked-On: https://jira01.devtools.intel.com/browse/<PROJ>-<NUMB>
## Several Tracked-On lines are allowed
## you can fully prefill your Tracked-On commits using environment variable:
## export CURRENT_TRACKED_ON=https://jira01.devtools.intel.com/browse/<PROJ>-<NUMB>
##
## Don't forget to add a JIRA URL to the Tracked-On metadata!
## https://wiki.ith.intel.com/display/CACTUS/Patch+requirements+check
## url should be in the form:
## Tracked-On: https://jira01.devtools.intel.com/browse/<PROJ>-<NUMB>
## Several Tracked-On lines are allowed
## you can fully prefill your Tracked-On commits using environment variable:
## export CURRENT_TRACKED_ON=https://jira01.devtools.intel.com/browse/<PROJ>-<NUMB>
##
---
 drivers/gpu/drm/i915/i915_drv.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/gpu/drm/i915/i915_drv.c b/drivers/gpu/drm/i915/i915_drv.c
index 2148636..b9e08a2 100644
--- a/drivers/gpu/drm/i915/i915_drv.c
+++ b/drivers/gpu/drm/i915/i915_drv.c
@@ -1015,6 +1015,9 @@ int i915_handle_hung_ring(struct drm_device *dev, uint32_t ringid)
 		DRM_DEBUG_TDR("Simulated gpu hang, rst stop_rings bits %08x\n",
 			(0x1 << ringid));
 		dev_priv->gpu_error.stop_rings &= ~(0x1 << ringid);
+		/* if all hangs are cleared, then clear the ALLOW_BAN/ERROR bits */
+		if ((dev_priv->gpu_error.stop_rings & ((1 << I915_NUM_RINGS) - 1)) == 0)
+			dev_priv->gpu_error.stop_rings = 0;
 	}
 
 	ret = intel_ring_disable(ring, current_context);
-- 
1.9.1

